### DISCUSSION TITLE
Add a clang-tidy file to ensure similar code style & conventions

### DISCUSSION THREAD

**Discussion by zjeffer** [Alexays/Waybar#2591] [Ideas]
Because this repo is getting contributions by lots of people, different code conventions are often mixed.

I propose we add a clang-tidy file to the repo, so everyone's IDE will let them know what code style to use. clang-format is already present, but I don't think it's enough. Adding a clang-tidy file would not force people to use the conventions, but would encourage them to.

We could fix readability issues like [else-after-return](https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html), and ensure everyone uses [the same casing for methods, classes, and variables](https://clang.llvm.org/extra/clang-tidy/checks/readability/identifier-naming.html) 

 'Modern' C++ stuff like using pass-by-value, using auto to avoid specifying the type twice in one line, or ensuring people use make_shared and make_unique instead of calling the smart pointer's constructor, etc... 

Here's the full list of clang-tidy checks we could add: https://clang.llvm.org/extra/clang-tidy/checks/list.html

**Comment by Syndelis** [comment]
Just wanted to leave a comment to show support, but I have nothing more to add; it's a great idea!

**Comment by zjeffer** [comment]
I'm currently using this file on my system, which seems to bring up quite a lot of issues:

```
Checks: >
  -*,
  bugprone-*
  misc-*,
  modernize-*,
  performance-*,
  portability-*,
  readability-*,
  -fuchsia-trailing-return,
  -readability-magic-numbers,
  -modernize-use-nodiscard,
  -modernize-use-trailing-return-type,
  -readability-braces-around-statements,
  -readability-redundant-access-specifiers,
  -readability-redundant-member-init,
  -readability-redundant-string-init,
  -readability-identifier-length
```

And then I add these for checking case:

```
CheckOptions:
  - { key: readability-identifier-naming.NamespaceCase,          value: lower_case }
  - { key: readability-identifier-naming.ClassCase,              value: CamelCase  }
  - { key: readability-identifier-naming.StructCase,             value: CamelCase  }
  - { key: readability-identifier-naming.FunctionCase,           value: CamelCase  }
  - { key: readability-identifier-naming.VariableCase,           value: camelBack  }
  - { key: readability-identifier-naming.PrivateMemberCase,      value: camelBack  }
  - { key: readability-identifier-naming.PrivateMemberPrefix,    value: m_         }
  - { key: readability-identifier-naming.EnumCase,               value: CamelCase  }
  - { key: readability-identifier-naming.EnumConstantCase,       value: UPPER_CASE }
  - { key: readability-identifier-naming.GlobalConstantCase,     value: UPPER_CASE }
```

For now enabling these casing checks report loads of problems in the files, because everyone seems to use different conventions. Maybe we can add just the Checks for now, until we can agree on the CheckOptions for casing? @Alexays what do you think?

**Comment by zjeffer** [comment]
Created a PR for the initial clang-tidy file: https://github.com/Alexays/Waybar/pull/2595

**Comment by zjeffer** [comment]
PR has been merged, closing...

**Comment by zjeffer** [comment]
@Alexays I noticed you disabled the clang-tidy GitHub Action, was it because of the variable naming? 
I think we could reenable it but remove the warnings for case styling? I would also remove the warning for function complexity, as that's also quite an annoying one that doesn't really help much. But I think the others have merit.
